### ISSUE TITLE
[MINVOKER-107] mvn deploy of SNAPSHOT failed with maven3 when invoker:run look for current build dependency

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-invoker-plugin#422]
**[Amertum](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=amertum)** opened **[MINVOKER-107](https://issues.apache.org/jira/browse/MINVOKER-107?redirect=false)** and commented

When I build http://code.google.com/p/mycila/source/browse/mycila-testing/trunk?r=1448 on revision 1448, the build failed. r1449 is a workaround.

the maven command is : mvn clean deploy
the build failed when running plugin invoker:run with maven3 BUT succeed with maven2
the failure is due of a missing artifact in the its local repository, the searched version is a SNAPSHOT, the present version is a timestamped SNAPSHOT

the dependency com.mycila.testing:mycila-testing-api:jar:2.6-SNAPSHOT is not found because the invoker:install copy the SNAPSHOT with the timestamp during deploy. And then after during the test, the dependency could not be found. The reason is because maven3 always deploy using a timestamped version.

`@see` https://cwiki.apache.org/MAVEN/maven-3x-compatibility-notes.html#Maven3.xCompatibilityNotes-NonuniqueSnapshotDeployments

! workaround
remove \<localRepositoryPath>${project.build.directory}/local-repo\</localRepositoryPath> from the invoker configuration

---

**Affects:** 1.5

**Attachments:**
- [MINVOKER-107.diff](https://issues.apache.org/jira/secure/attachment/12716656/MINVOKER-107.diff) (_702 bytes_)

6 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Tony Chemit](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tchemit)** commented

Here is patch which fix the problem.

It works fine (all its are ok except the one which was already broken before the patch).

I don't know exatcly the ArtifactInstaller api with also the ArtifactTransformation one.

Hope somebody can have a look at it and correct the patch if required.

Thanks.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Tony, could you add an unit-test or IT to verify this and to prevent regression?

**Comment by jira-importer** [comment]
**[Yegor Bugayenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fazend)** commented

What is really a workaround at the moment?

**Comment by jira-importer** [comment]
**[Tony Chemit](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tchemit)** commented

IMHO mine is an acceptable one, ludovic workaround is not acceptable since you can not ask maven user to spoil his local repository with its (but you can do it for yourself).

My ultimate workaround is to NOT deploy for a snapshot with running its at the same time :(

**Comment by jira-importer** [comment]
**[Tony Chemit](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tchemit)** commented

Robert :

> Tony, could you add an unit-test or IT to verify this and to prevent regression?

I found a project in the haus which predocue this problem : the webstart-maven-plugin.

svn : https://svn.codehaus.org/mojo/trunk/mojo/webstart/webstart-maven-plugin

Just try a mvn clean deploy and it will failed :(

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

ok, got it. Somehow the `clean` seems to cause trouble: I've tried `mvn deploy` twice without any problem. To be continued.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Tony, I've tried to reproduce this issue and somehow I can't anymore. Are you still having this issue?

**Comment by jira-importer** [comment]
**[Tony Chemit](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tchemit)** commented

Robert, Same for me, I can no more reproduce it on the webstart project. I will try with other projects I know it fails

**Comment by jira-importer** [comment]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dawid.weiss)** commented

This issue is a bug and is repeatable for me. The invoker plugin installs shapshot versions of artefacts but this happens ONLY if running:

```
mvn clean deploy
```

and NOT if running:

```
mvn clean verify
```

So it is the deployment of artefacts that somehow renders -SNAPSHOT versions to concrete numbered revisions. Build logs from our Bamboo where this is apparent:

Passing build (mvn clean verify):
http://builds.carrot2.org/browse/RNDTEST-MASTER-JOB1-109/log

Failing build (mvn clean deploy):
http://builds.carrot2.org/browse/RNDTEST-MASTER-JOB1-108/log

Passing build's invoker install section:

```
22-Jan-2012 11:31:03	[INFO] --- maven-invoker-plugin:1.5:install (integration-test) @ junit4-maven-plugin-tests ---
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/runner/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-runner/0.0.5-SNAPSHOT/randomizedtesting-runner-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/runner/target/randomizedtesting-runner-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-runner/0.0.5-SNAPSHOT/randomizedtesting-runner-0.0.5-SNAPSHOT.jar
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-parent/0.0.5-SNAPSHOT/randomizedtesting-parent-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin/0.0.5-SNAPSHOT/junit4-maven-plugin-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin/target/junit4-maven-plugin-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin/0.0.5-SNAPSHOT/junit4-maven-plugin-0.0.5-SNAPSHOT.jar
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-ant/ant-junit4/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/ant-junit4/0.0.5-SNAPSHOT/ant-junit4-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-ant/ant-junit4/target/ant-junit4-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/ant-junit4/0.0.5-SNAPSHOT/ant-junit4-0.0.5-SNAPSHOT.jar
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin-tests/0.0.5-SNAPSHOT/junit4-maven-plugin-tests-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:31:04	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/junit4-maven-plugin-tests-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin-tests/0.0.5-SNAPSHOT/junit4-maven-plugin-tests-0.0.5-SNAPSHOT.jar
```

Failing build's invoker install section:

```
22-Jan-2012 11:26:32	[INFO] --- maven-invoker-plugin:1.5:install (integration-test) @ junit4-maven-plugin-tests ---
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/runner/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-runner/0.0.5-SNAPSHOT/randomizedtesting-runner-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/runner/target/randomizedtesting-runner-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-runner/0.0.5-SNAPSHOT/randomizedtesting-runner-0.0.5-20120122.102516-54.jar
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/runner/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-runner/0.0.5-SNAPSHOT/randomizedtesting-runner-0.0.5-20120122.102516-54.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/randomizedtesting-parent/0.0.5-SNAPSHOT/randomizedtesting-parent-0.0.5-20120122.102443-55.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin/0.0.5-SNAPSHOT/junit4-maven-plugin-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin/target/junit4-maven-plugin-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin/0.0.5-SNAPSHOT/junit4-maven-plugin-0.0.5-20120122.102625-6.jar
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin/0.0.5-SNAPSHOT/junit4-maven-plugin-0.0.5-20120122.102625-6.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-ant/ant-junit4/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/ant-junit4/0.0.5-SNAPSHOT/ant-junit4-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-ant/ant-junit4/target/ant-junit4-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/ant-junit4/0.0.5-SNAPSHOT/ant-junit4-0.0.5-20120122.102614-54.jar
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-ant/ant-junit4/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/ant-junit4/0.0.5-SNAPSHOT/ant-junit4-0.0.5-20120122.102614-54.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/pom.xml to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin-tests/0.0.5-SNAPSHOT/junit4-maven-plugin-tests-0.0.5-SNAPSHOT.pom
22-Jan-2012 11:26:33	[INFO] Installing /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/junit4-maven-plugin-tests-0.0.5-SNAPSHOT.jar to /home/bamboo/bamboo3-work/xml-data/build-dir/RNDTEST-MASTER-JOB1/integration-maven/junit4-maven-plugin-tests/target/local-repo/com/carrotsearch/randomizedtesting/junit4-maven-plugin-tests/0.0.5-SNAPSHOT/junit4-maven-plugin-tests-0.0.5-SNAPSHOT.jar
```

**Comment by jira-importer** [comment]
**[Dawid Weiss](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dawid.weiss)** commented

Same with Maven 3.0.3.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Just for the record: I am able to regularly reproduce this bug in combination of Maven 3.0.3 and invoking `deploy` goal (I was baffled why it passes on me locally, and failing on CI: simply, I was doing as usual `mvn clean install` and CI was doing `mvn clean deploy`). This was stated above too by David.

OTOH, Maven 3.0.4 for me locally, does NOT triggers this bug, and the `invoker:install` step goes just fine with `deploy` goal invoked on CLI.

So for me, workaround is to upgrade CI to Maven 3.0.4 (as for me locally `mvn clean deploy` + Maven 3.0.4 does not fails, in contrary as doing same with Maven 3.0.3, which does).

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I'd like to close this as "not a bug", the root cause seems to be in Maven and is fixed in version 3.0.4
